File: ios/SceneGo/Speech/SceneGoSpeechRecognizer.m
#import "SceneGoSpeechRecognizer.h"
#import <Speech/Speech.h>
#import <AVFoundation/AVFoundation.h>

@interface SceneGoSpeechRecognizer () <SFSpeechRecognizerDelegate>
@property (nonatomic, strong) SFSpeechRecognizer *speechRecognizer;
@property (nonatomic, strong) SFSpeechAudioBufferRecognitionRequest *recognitionRequest;
@property (nonatomic, strong) SFSpeechRecognitionTask *recognitionTask;
@property (nonatomic, strong) AVAudioEngine *audioEngine;
@property (nonatomic, assign) BOOL hasListeners;
@end

@implementation SceneGoSpeechRecognizer

RCT_EXPORT_MODULE(SceneGoSpeechRecognizer);

+ (BOOL)requiresMainQueueSetup {
    return YES;
}

- (instancetype)init {
    self = [super init];
    if (self) {
        _audioEngine = [[AVAudioEngine alloc] init];
        // recognizer 在每次 startListening 时按传入 locale 构建（含可用性降级）
        _speechRecognizer = [self buildRecognizerForLocale:@"zh-CN"];
    }
    return self;
}

/** 构建指定 locale 的识别器；不可用时自动降级到同语言前缀或 en-US */
- (SFSpeechRecognizer *)buildRecognizerForLocale:(NSString *)localeId {
    NSArray<NSLocale *> *supported = [SFSpeechRecognizer supportedLocales];
    if (supported.count == 0) return nil;

    // 精确匹配
    for (NSLocale *loc in supported) {
        if ([[loc localeIdentifier] isEqualToString:localeId]) {
            return [[SFSpeechRecognizer alloc] initWithLocale:loc];
        }
    }
    // 同语言前缀（如 zh-CN → zh-Hans / zh-TW）
    NSString *langPrefix = [[localeId componentsSeparatedByString:@"-"] firstObject];
    for (NSLocale *loc in supported) {
        if ([[loc localeIdentifier] hasPrefix:langPrefix]) {
            NSLog(@"[Speech] %@ 不可用，降级为 %@", localeId, loc.localeIdentifier);
            return [[SFSpeechRecognizer alloc] initWithLocale:loc];
        }
    }
    // 英文兑底
    for (NSLocale *loc in supported) {
        if ([[loc localeIdentifier] hasPrefix:@"en"]) {
            NSLog(@"[Speech] %@ 不可用，降级为 %@", localeId, loc.localeIdentifier);
            return [[SFSpeechRecognizer alloc] initWithLocale:loc];
        }
    }
    NSLog(@"[Speech] %@ 不可用，使用 %@", localeId, supported.firstObject.localeIdentifier);
    return [[SFSpeechRecognizer alloc] initWithLocale:supported.firstObject];
}

- (NSArray<NSString *> *)supportedEvents {
    return @[@"onSpeechResult", @"onSpeechError"];
}

- (void)startObserving {
    _hasListeners = YES;
}

- (void)stopObserving {
    _hasListeners = NO;
}

RCT_EXPORT_METHOD(startListening:(NSString *)localeStr
                  resolver:(RCTPromiseResolveBlock)resolve
                  rejecter:(RCTPromiseRejectBlock)reject)
{
    [SFSpeechRecognizer requestAuthorization:^(SFSpeechRecognizerAuthorizationStatus status) {
        dispatch_async(dispatch_get_main_queue(), ^{
            if (status != SFSpeechRecognizerAuthorizationStatusAuthorized) {
                reject(@"auth_denied", @"Speech recognition permission denied", nil);
                return;
            }

            // 按传入 locale 构建识别器（内部处理可用性降级）
            SFSpeechRecognizer *recognizer = [self buildRecognizerForLocale:localeStr.length > 0 ? localeStr : @"zh-CN"];
            if (!recognizer) {
                reject(@"locale_unavailable", @"No speech recognizer available for this device", nil);
                return;
            }
            recognizer.delegate = self;
            self.speechRecognizer = recognizer;
            [self startAudioRecordingWithRecognizer:recognizer resolver:resolve rejecter:reject];
        });
    }];
}

- (void)startAudioRecordingWithRecognizer:(SFSpeechRecognizer *)recognizer
                                 resolver:(RCTPromiseResolveBlock)resolve
                                 rejecter:(RCTPromiseRejectBlock)reject
{
    if (_recognitionTask) {
        [_recognitionTask cancel];
        _recognitionTask = nil;
    }

    AVAudioSession *audioSession = [AVAudioSession sharedInstance];
    NSError *error = nil;
    [audioSession setCategory:AVAudioSessionCategoryRecord mode:AVAudioSessionModeMeasurement options:AVAudioSessionCategoryOptionDuckOthers error:&error];
    [audioSession setActive:YES withOptions:AVAudioSessionSetActiveOptionNotifyOthersOnDeactivation error:&error];

    _recognitionRequest = [[SFSpeechAudioBufferRecognitionRequest alloc] init];
    AVAudioInputNode *inputNode = _audioEngine.inputNode;

    if (!_recognitionRequest || !inputNode) {
        reject(@"init_error", @"Unable to create recognition request or input node", nil);
        return;
    }

    _recognitionRequest.shouldReportPartialResults = YES;

    __weak typeof(self) weakSelf = self;
    _recognitionTask = [recognizer recognitionTaskWithRequest:_recognitionRequest resultHandler:^(SFSpeechRecognitionResult * _Nullable result, NSError * _Nullable error) {
        typeof(self) strongSelf = weakSelf;
        if (!strongSelf) return;

        if (result) {
            NSString *transcript = result.bestTranscription.formattedString;
            if (strongSelf.hasListeners) {
                [strongSelf sendEventWithName:@"onSpeechResult" body:@{
                    @"transcript": transcript,
                    @"isFinal": @(result.isFinal)
                }];
            }
        }

        if (error) {
            // 错误事件上报（权限受限/网络不可达/识别引擎故障等）
            NSLog(@"[Speech] recognition error: %@", error.localizedDescription);
            if (strongSelf.hasListeners) {
                [strongSelf sendEventWithName:@"onSpeechError" body:@{ @"message": error.localizedDescription ?: @"unknown" }];
            }
        }

        if (error || (result && result.isFinal)) {
            [strongSelf.audioEngine stop];
            [inputNode removeTapOnBus:0];
            strongSelf.recognitionRequest = nil;
            strongSelf.recognitionTask = nil;
        }
    }];

    AVAudioFormat *recordingFormat = [inputNode outputFormatForBus:0];
    [inputNode installTapOnBus:0 bufferSize:1024 format:recordingFormat block:^(AVAudioPCMBuffer * _Nonnull buffer, AVAudioTime * _Nonnull when) {
        [weakSelf.recognitionRequest appendAudioPCMBuffer:buffer];
    }];

    [_audioEngine prepare];
    [_audioEngine startAndReturnError:&error];

    if (error) {
        reject(@"engine_error", error.localizedDescription, error);
    } else {
        resolve(@YES);
    }
}

RCT_EXPORT_METHOD(stopListening:(RCTPromiseResolveBlock)resolve rejecter:(RCTPromiseRejectBlock)reject)
{
    dispatch_async(dispatch_get_main_queue(), ^{
        if (self.audioEngine.isRunning) {
            [self.audioEngine stop];
            [self.recognitionRequest endAudio];
        }
        if (self.recognitionTask) {
            [self.recognitionTask cancel];
            self.recognitionTask = nil;
        }
        resolve(@YES);
    });
}

@end

File: modules/scenego-speech/ios/SceneGoSpeechRecognizer.swift
import ExpoModulesCore
import Speech
import AVFoundation

public class SceneGoSpeechRecognizer: Module {
  private var speechRecognizer: SFSpeechRecognizer?
  private var recognitionRequest: SFSpeechAudioBufferRecognitionRequest?
  private var recognitionTask: SFSpeechRecognitionTask?
  private let audioEngine = AVAudioEngine()
  // 授权回调期间收到 stop 时置位，授权完成后放弃启动（消除快速连点导致的双任务竞争）
  private var stopRequested = false

  /// 发送事件：旧架构 (Paper) 下必须经 RCT 事件桥 (EXReactNativeEventEmitter) 才能到达 JS
  /// （LegacyEventEmitterCompat 的 JSI 路径在旧架构 runtime 不可用，事件会被静默丢弃）
  private func emitEvent(_ name: String, _ body: [String: Any]) {
    if let serviceProtocol = NSProtocolFromString("EXEventEmitterService"),
       let legacyEmitter = (appContext?.legacyModule(implementing: serviceProtocol) as NSObject?) {
      legacyEmitter.perform(NSSelectorFromString("sendEventWithName:body:"), with: name, with: body as NSDictionary)
    } else {
      // 新架构 fallback：JSI 事件
      sendEvent(name, body)
    }
  }

  public func definition() -> ModuleDefinition {
    Name("SceneGoSpeechRecognizer")

    Events("onSpeechResult", "onSpeechError")

    AsyncFunction("startListening") { (promise: Promise) in
      self.startListening(locale: "zh-CN", promise: promise)
    }.runOnQueue(.main)

    AsyncFunction("stopListening") { (promise: Promise) in
      self.stopListening()
      promise.resolve(true)
    }.runOnQueue(.main)
  }

  private func startListening(locale: String, promise: Promise) {
    stopRequested = false
    SFSpeechRecognizer.requestAuthorization { status in
      DispatchQueue.main.async {
        // 授权窗口内用户已点关闭：放弃本次启动，避免 stop 后录音仍在后台运行
        guard !self.stopRequested else {
          promise.resolve(false)
          return
        }
        guard status == .authorized else {
          promise.reject("auth_denied", "Speech recognition permission denied")
          return
        }
        guard let recognizer = self.buildRecognizer(forLocale: locale) else {
          promise.reject("locale_unavailable", "No speech recognizer available for this device")
          return
        }
        self.speechRecognizer = recognizer
        self.beginRecording(with: recognizer, promise: promise)
      }
    }
  }

  /// 构建指定 locale 的识别器；不可用时自动降级到同语言前缀或 en-US
  private func buildRecognizer(forLocale localeId: String) -> SFSpeechRecognizer? {
    let supported = SFSpeechRecognizer.supportedLocales()
    guard !supported.isEmpty else { return nil }

    // 精确匹配
    if let exact = supported.first(where: { $0.identifier == localeId }) {
      return SFSpeechRecognizer(locale: exact)
    }
    // 同语言前缀（如 zh-CN → zh-Hans / zh-TW）
    let langPrefix = localeId.split(separator: "-").first.map(String.init) ?? ""
    if let sameLang = supported.first(where: { $0.identifier.hasPrefix(langPrefix) }) {
      NSLog("[Speech] %@ 不可用，降级为 %@", localeId, sameLang.identifier)
      return SFSpeechRecognizer(locale: sameLang)
    }
    // 英文兜底
    if let en = supported.first(where: { $0.identifier.hasPrefix("en") }) {
      NSLog("[Speech] %@ 不可用，降级为 %@", localeId, en.identifier)
      return SFSpeechRecognizer(locale: en)
    }
    if let first = supported.first {
      NSLog("[Speech] %@ 不可用，使用 %@", localeId, first.identifier)
      return SFSpeechRecognizer(locale: first)
    }
    return nil
  }

  private func beginRecording(with recognizer: SFSpeechRecognizer, promise: Promise) {
    if let task = recognitionTask {
      task.cancel()
      recognitionTask = nil
    }

    let audioSession = AVAudioSession.sharedInstance()
    do {
      try audioSession.setCategory(.record, mode: .measurement, options: .duckOthers)
      try audioSession.setActive(true, options: .notifyOthersOnDeactivation)
    } catch {
      promise.reject("audio_session_error", error.localizedDescription)
      return
    }

    let request = SFSpeechAudioBufferRecognitionRequest()
    request.shouldReportPartialResults = true
    recognitionRequest = request

    let inputNode = audioEngine.inputNode
    if inputNode.inputFormat(forBus: 0).sampleRate == 0 {
      // 无输入设备（模拟器未授权麦克风等）
      promise.reject("no_input_device", "No audio input device available")
      return
    }

    recognitionTask = recognizer.recognitionTask(with: request) { [weak self] result, error in
      guard let self else { return }

      if let result {
        let transcript = result.bestTranscription.formattedString
        self.emitEvent("onSpeechResult", [
          "transcript": transcript,
          "isFinal": result.isFinal,
        ])
      }

      if let error {
        // cancel 是用户主动停止触发的正常错误（NSUserCancelledError），不视为异常，不上报
        let nsError = error as NSError
        if nsError.code != NSUserCancelledError {
          NSLog("[Speech] recognition error: %@", error.localizedDescription)
          self.emitEvent("onSpeechError", ["message": error.localizedDescription])
        }
      }

      if error != nil || (result?.isFinal ?? false) {
        self.audioEngine.stop()
        inputNode.removeTap(onBus: 0)
        self.recognitionRequest = nil
        self.recognitionTask = nil
      }
    }

    let recordingFormat = inputNode.outputFormat(forBus: 0)
    inputNode.installTap(onBus: 0, bufferSize: 1024, format: recordingFormat) { [weak self] buffer, _ in
      self?.recognitionRequest?.append(buffer)
    }

    audioEngine.prepare()
    do {
      try audioEngine.start()
      promise.resolve(true)
    } catch {
      promise.reject("engine_error", error.localizedDescription)
    }
  }

  private func stopListening() {
    stopRequested = true
    if audioEngine.isRunning {
      audioEngine.stop()
      recognitionRequest?.endAudio()
    }
    recognitionTask?.cancel()
    recognitionTask = nil
    // 释放录音会话，避免持续占用麦克风
    try? AVAudioSession.sharedInstance().setActive(false, options: .notifyOthersOnDeactivation)
  }
}
